custom/icds/translations/integrations/transifex.py

```diff
@@ -0,0 +1,200 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+import polib
+import datetime
+import os
+
+from django.utils.translation import ugettext as _
+from django.conf import settings
+
+from collections import namedtuple, OrderedDict
+from corehq.apps.app_manager import id_strings
+from custom.icds.translations.integrations.client import TransifexApiClient
+
+Translation = namedtuple('Translation', 'key translation occurrences')
+Unique_ID = namedtuple('UniqueID', 'type id')
+
+
+class Transifex():
+    def __init__(self, domain, app_id, source_lang, target_lang=None, version=None, lang_prefix='default_'):
+        """
+        :param domain: domain name
+        :param app_id: id of the app to be used
+        :param source_lang: source lang code like en
+        :param target_lang: target lang code like hin
+        :param version: version of the app like 10475
+        :param lang_prefix: prefix if other than "default_"
+        """
+        self.app_id = app_id
+        self.domain = domain
+        self.version = int(version)
+        self.source_lang = source_lang
+        self.target_lang = target_lang
+        self.lang_prefix = lang_prefix
+        if not self.target_lang:
+            self.target_lang = self.source_lang
+        self.app_id_to_build = None
+        self.translations = OrderedDict()
+        self.headers = dict()  # headers for each sheet name
+        self.sheet_name_to_module_or_form_type_and_id = dict()
+        self.generated_files = list()
```

Comment:
what do you mean by a single resource here?

Reply:
I mean one sheet/file - one set of translations, probably corresponding to a single form or module.

Right now, `translations`, `headers`, `sheet_name_to_module_or_form_type_and_id`, and `generated_files` each store a collection of data broken up by file.  For example, `self.translations` is a dict mapping file names to the translations for that file.

Instead, you could have a single property called `self.resources` (or whatever), which is a list of `Resource` objects (or a map of filenames to `Resource` objects).  Then each `Resource` object would manage it's own translations, headers, module/form type and id, and so forth.

This way, the broad strokes would happen in this parent class, and the child `Resource` class instances would handle the detail work, but they'd have a smaller set of state - only that corresponding to a single form or module.

Reply:
hmm, all the translations and headers are received at once and `sheet_name_to_module_or_form_type_and_id` would also be generated at once from the first sheet. Its then essentially this method `_get_translation_for_sheet` that generates translation for a resource as needed for a po file, so it would be essentially be just this method in `Resource`. Will try that out to see if that makes sense to extract out.